cupy/cuda/texture.pyx

```diff
@@ -0,0 +1,385 @@
+
+        # get src
+        if src is self:
+            # Important: cannot convert from src.ptr!
+            param.srcArray = <Array>(self.ptr)
+            param.extent = runtime.make_Extent(self.width, self.height,
+                                               self.depth)
+        else:
+            width = src.shape[-1]
+            if src.ndim >= 2:
+                height = src.shape[-2]
+            else:
+                height = 1  # same "one-stride" trick here
+
+            if isinstance(src, ndarray):
+                ptr = src.data.ptr
+            else:  # numpy.ndarray
+                ptr = src.ctypes.data
+
+            srcPitchedPtr = runtime.make_PitchedPtr(
+                ptr, width*src.dtype.itemsize, width, height)
+            param.srcPtr = srcPitchedPtr
+
+        # get dst
+        if dst is self:
+            # Important: cannot convert from dst.ptr!
+            param.dstArray = <Array>(self.ptr)
+            param.extent = runtime.make_Extent(self.width, self.height,
+                                               self.depth)
+        else:
+            width = dst.shape[-1]
+            if dst.ndim >= 2:
+                height = dst.shape[-2]
+            else:
+                height = 1  # same "one-stride" trick here
+
+            if isinstance(dst, ndarray):
+                ptr = dst.data.ptr
+            else:  # numpy.ndarray
+                ptr = dst.ctypes.data
+
+            dstPitchedPtr = runtime.make_PitchedPtr(
+                ptr, width*dst.dtype.itemsize, width, height)
+            param.dstPtr = dstPitchedPtr
+
+        return <void*>param
+
+    def _prepare_copy(self, arr, stream, direction):
+        # sanity checks:
+        # - check shape
+        if self.ndim == 3:
+            if arr.shape != (self.depth, self.height, self.width):
+                raise ValueError("shape mismatch")
+        elif self.ndim == 2:
+            if arr.shape != (self.height, self.width):
+                raise ValueError("shape mismatch")
+        else:  # ndim = 1
+            if arr.shape != (self.width,):
+                raise ValueError("shape mismatch")
+
+        # - check dtype
+        # TODO(leofang): we should also check channel bit size vs dtype
+        # itemsize, but can we assume it's always single channel?
+        ch_kind = self.desc.get_channel_format()['f']
+        if ch_kind == runtime.cudaChannelFormatKindSigned:
+            if arr.dtype not in (numpy.int8, numpy.int16, numpy.int32):
+                raise ValueError("dtype mismatch")
+        elif ch_kind == runtime.cudaChannelFormatKindUnsigned:
+            if arr.dtype not in (numpy.uint8, numpy.uint16, numpy.uint32):
+                raise ValueError("dtype mismatch")
+        elif ch_kind == runtime.cudaChannelFormatKindFloat:
+            if arr.dtype not in (numpy.float16, numpy.float32):
+                raise ValueError("dtype mismatch")
+        else:
+            raise ValueError("dtype not supported")
```

Comment:
Here, I tend to not check the bit sizes in x, y, z, and w of the `ChannelFormatDescriptor`, because I am not sure how users would fill in the texture memory and reinterpret the data when calling `RawKernel` (say reading two floats to make up a complex). Since this is a quite advanced feature, I say let's do minimal checks and let users be responsible for it.

Reply:
I think it's fine as we are only exposing the API.